### PR TITLE
Added support for DynamoDB Continuous Backups.

### LINKED ISSUE #4661: Let Teleport to initialize DynamoDB tables with auto-scaling enabled
We need to be able to initialize DynamoDB tables with autoscalling enabled. The additional settings for the auto scaling policies may come from the config file. @webvictim can provide more information on the extra parameters. These extra parameters do not need to be documented and should be considered internal. 

@russjones would be nice to make it part of 5.0.

### PR DESCRIPTION
**Pull request by russjones**
**Description**

Added support for DynamoDB Continuous Backups and Auto Scaling.

To use, set the following variables in Teleport file configuration.

```yaml
teleport:
  storage:
    type: "dynamodb"
    table_name: "rusty-backend"
    region: "us-east-1"
    [...]

    # continuous_backups is used to enable or disable continuous backups.
    continuous_backups: [true|false]

    # auto_scaling is used to enable (and define settings for) or disable auto scaling.
    auto_scaling:  [true|false]
    read_min_capacity: int
    read_max_capacity: int
    read_target_value: float
    write_min_capacity: int
    write_max_capacity: int
    write_target_value: float
```

**Related Issues**

Fixes https://github.com/gravitational/teleport/issues/4661

### CODE REVIEW

**Comment by alex-kovoy** [comment]
@russjones can we add parameters for auto-scaling as well?

**Comment by alex-kovoy** [comment]
@russjones https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/AutoScaling.Console.html#AutoScaling.Console.NewTable

**Comment by russjones** [comment]
@alex-kovoy @webvictim Auto scaling has a lot of configuration, what do you want to specify?

Looks like you can configure both both read and write capacity and for each one target utilization, minimum and maximum provisioned capacity.

I might also need to know the linked IAM role name?

File: lib/backend/dynamo/dynamodbbk.go [hunk 1]
```diff
@@ -65,6 +65,8 @@ type Config struct {
 	PollStreamPeriod time.Duration `json:"poll_stream_period,omitempty"`
 	// RetryPeriod is a period between dynamo backend retries on failures
 	RetryPeriod time.Duration `json:"retry_period"`
+	// EnableContinuousBackups enabled PITR (Point-In-Time) Recovery.
```
Comment (webvictim) [inline lib/backend/dynamo/dynamodbbk.go:68]:
```suggestion
	// EnableContinuousBackups enables PITR (Point-In-Time Recovery).
```

File: lib/backend/dynamo/dynamodbbk_test.go [hunk 2]
```diff
@@ -101,3 +103,37 @@ func (s *DynamoDBSuite) TestWatchersClose(c *check.C) {
...
+// TestContinuousBackups verifies that the continuous backup state is updated
+// upon start of the DynamoDB backend.
+func (s *DynamoDBSuite) TestContinuousBackups(c *check.C) {
+	var tests = []struct {
+		enabled bool
+		desc    check.CommentInterface
+	}{
+		{
+			enabled: true,
+			desc:    check.Commentf("enable continuous backups"),
```
Comment (webvictim) [inline lib/backend/dynamo/dynamodbbk_test.go:116]:
```suggestion
			desc:    check.Commentf("enabled continuous backups"),
```

**Comment by webvictim** [comment]
#### This is the way that our reference Terraform currently provisions the DynamoDB cluster storage table:

2020-10-30-113859__794x852__maim

I think that to start with we should add an option to provision both tables with autoscaling, using the same default parameters as shown above:

Target utilization: 50%
Minimum: 5 units
Maximum: 100 units

I'm not sure how we should handle the ARN for autoscaling - here's the logic for creating the role in Terraform: https://github.com/gravitational/teleport/blob/master/examples/aws/terraform/ha-autoscale-cluster/dynamo.tf#L102-L140

As the intention is for this to be used internally, maybe we should just provide the ability to specify the role ARN to assume for autoscaling? (and assume that this role is provisioned with correct parameters externally, rather than assuming that Teleport will be running with sufficient IAM permissions to create a role)

I think that ideally we'd expose all of these settings via config flags, but I suspect that would probably require considerably more testing? Maybe we just use these as default limits when autoscaling is enabled for now. What do you think @alex-kovoy @pierrebeaucamp?

**Comment by benarent** [comment]
> Minimum: 5 units

^ I have a feeling from support that it's best to bump this to `20 units`, I can't find the specific ticket but 5 is the lowest and I know customers have ran into issues with it. https://github.com/gravitational/teleport/issues/4166

**Comment by alex-kovoy** [comment]
Lets use the default values for all tables and pass the role ARN name via configuration. This should be good for initial start. @klizhentas what do you think?

**Comment by russjones** [comment]
@alex-kovoy @webvictim @pierrebeaucamp I might be reading the docs wrong, but I think auto scaling information comes from some policy?

Let me see how to attach a policy to the table and get back to you folks. It might be something you just define yourself and Teleport picks up a path and loads the polciy?

File: lib/backend/dynamo/dynamodbbk.go [hunk 3]
```diff
@@ -822,6 +828,52 @@ func (b *Backend) getKey(ctx context.Context, key []byte) (*record, error) {
...
+			PointInTimeRecoveryEnabled: aws.Bool(b.Config.EnableContinuousBackups),
+		},
+		TableName: aws.String(b.Tablename),
+	})
+	if err != nil {
+		return convertError(err)
+	}
+
+	// Make sure the state Dynamo is now in matches the expected state.
+	status, err := b.getContinuousBackups(ctx)
```
Comment (a-palchikov) [inline lib/backend/dynamo/dynamodbbk.go:865]: Why is this necessary? I'm not familiar with dynamodb but from what I read about `UpdateContinuousBackups` API it seems to ascertain that after a successful invocation response contains the `ContinuousBackDescription` which reflects the update:
```
...A successful UpdateContinuousBackups call returns the current ContinuousBackupsDescription.
Continuous backups are ENABLED on all tables at table creation. If point in time recovery is enabled, PointInTimeRecoveryStatus will be set to ENABLED...
```
Is it b/c it was found to return the result inconsistently or for other reason?

**Review by a-palchikov — APPROVED** [review]

File: lib/backend/dynamo/dynamodbbk.go (hunk 3 above)
Comment (russjones) [reply lib/backend/dynamo/dynamodbbk.go:865]: Yeah, `ContinuousBackupsStatus` in the response always appears to have a invalid response. I could assume no error meant the call was successful, but since I had to add `getContinuousBackups` for tests, figured might as well verify the property was set correctly.

File: lib/backend/dynamo/dynamodbbk.go [hunk 4]
```diff
@@ -822,6 +828,52 @@ func (b *Backend) getKey(ctx context.Context, key []byte) (*record, error) {
...
+		return false, convertError(err)
+	}
+
+	switch *resp.ContinuousBackupsDescription.PointInTimeRecoveryDescription.PointInTimeRecoveryStatus {
+	case string(dynamodb.ContinuousBackupsStatusEnabled):
+		return true, nil
+	case string(dynamodb.ContinuousBackupsStatusDisabled):
+		return false, nil
+	default:
+		return false, trace.BadParameter("dynamo return unknown state for continuous backups: %v",
```
Comment (webvictim) [inline lib/backend/dynamo/dynamodbbk.go:846]:
```suggestion
		return false, trace.BadParameter("dynamo returned unknown state for continuous backups: %v",
```

**Comment by webvictim** [comment]
This is how Terraform is doing it: https://github.com/gravitational/teleport/blob/master/examples/aws/terraform/ha-autoscale-cluster/dynamo.tf#L167-L215

Looks like behind the scenes there's a read policy and a write policy which have targets attached. I'm not 100% sure how it's actually implemented.

File: lib/backend/dynamo/dynamodbbk.go (hunk 3 above)
Comment (russjones) [reply lib/backend/dynamo/dynamodbbk.go]: @a-palchikov Actually I went with your suggestion, if an error is not returned I assume the operation was successful. I test for correctness in a test.

**Comment by russjones** [comment]
@alex-kovoy @webvictim @pierrebeaucamp Take a look, added support for auto scaling.

**Review by webvictim — APPROVED** [review]
LGTM

File: lib/backend/dynamo/dynamodbbk.go [hunk 5]
```diff
@@ -65,15 +65,35 @@ type Config struct {
 	PollStreamPeriod time.Duration `json:"poll_stream_period,omitempty"`
 	// RetryPeriod is a period between dynamo backend retries on failures
 	RetryPeriod time.Duration `json:"retry_period"`
+
+	// EnableContinuousBackups is used to enable (or disable) PITR (Point-In-Time Recovery).
+	EnableContinuousBackups bool `json:"continuous_backups"`
+
+	// EnableAutoScaling is used to enable (or disable) auto scaling policy.
+	EnableAutoScaling bool `json:"auto_scaling"`
```
Comment (r0mant) [inline lib/backend/dynamo/dynamodbbk.go:73]: Nit: All of these new fields are presumably optional, so I would consider adding `omitempty` to them.

**Review by r0mant — APPROVED** [review]

File: lib/backend/dynamo/dynamodbbk.go (hunk 5 above)
Comment (russjones) [reply lib/backend/dynamo/dynamodbbk.go]: Done.
